### PR TITLE
[JENKINS-27120] Adding Workflow support for JaCoCo publisher

### PR DESCRIPTION
**Pull request by lordofthejars**
There is no test since testing JaCoCo plugin in unit test it goes to a chicken-egg problem. JaCoCo needs a classes directory to copy them to its internal structure, so the first idea is to copy those classes before executing the job, but then Jenkins complains that this job has already created a directory and it is not executed. I have tested manually and it works and a test should be added to ATH.

In fact if you look at all JaCoCoPublisher tests that were already created you'll notice that they do nothing or are commented, probably because of the same problem I faced during Workflow integration.

Issue: https://issues.jenkins-ci.org/browse/JENKINS-27120

@reviewbybees

### CODE REVIEW

**Comment by recena** [comment]
@lordofthejars :+1: to review the coding style. Probably you will receive the common message "Unrelated changes", but...

**Comment by lordofthejars** [comment]
Well pom.xml file originally was formatted with tabs, white spaces (2 and 4)

File: pom.xml [hunk 1]
```diff
@@ -22,495 +22,536 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
...
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-settings</artifactId>
+            <version>3.3.9</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.codehaus.plexus</groupId>
+            <artifactId>plexus-utils</artifactId>
+            <version>3.0.15</version>
```
Comment (jtnord) [inline pom.xml]: no scope?

File: pom.xml [hunk 2]
```diff
@@ -22,495 +22,536 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 -->
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (jtnord) [inline pom.xml]: the number of unrelated changes in this file is excessive.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 3]
```diff
@@ -33,9 +36,9 @@
  * @author Jonathan Fuerth
  * @author Ognjen Bubalo
  */
-public final class JacocoBuildAction extends CoverageObject<JacocoBuildAction> implements HealthReportingAction, StaplerProxy, Serializable {
+public final class JacocoBuildAction extends CoverageObject<JacocoBuildAction> implements HealthReportingAction, StaplerProxy, Serializable, RunAction2 {
 
-	public final AbstractBuild<?,?> owner;
+	public transient Run<?,?> owner;
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: this is a backwards incompatable change.  Not that I am trying to suggest a BCT transform

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 4]
```diff
@@ -297,12 +293,12 @@ public JacocoBuildAction getPreviousResult() {
 	 * @throws IOException
 	 *      if failed to parse the file.
 	 */
-	public static JacocoBuildAction load(AbstractBuild<?,?> owner, Rule rule, JacocoHealthReportThresholds thresholds, BuildListener listener, JacocoReportDir layout, String[] includes, String[] excludes) throws IOException {
+	public static JacocoBuildAction load(Run<?,?> owner, JacocoHealthReportThresholds thresholds, TaskListener listener, JacocoReportDir layout, String[] includes, String[] excludes) throws IOException {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: API breaking change.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 5]
```diff
@@ -359,4 +355,12 @@ public final PrintStream getLogger() {
 	    // does not run the construct and thus leaves the transient variables empty
 	    return System.out;
 	}
+
+	public void onAttached(Run<?, ?> run) {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: `@Override`

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 6]
```diff
@@ -359,4 +355,12 @@ public final PrintStream getLogger() {
...
+		this.owner = run;
+	}
+
+	public void onLoad(Run<?, ?> run) {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: `@Override`

File: pom.xml (hunk 1 above)
Comment (lordofthejars) [inline pom.xml]: I don't know it was in the original pom, I have not introduced this change.

File: pom.xml (hunk 2 above)
Comment (lordofthejars) [inline pom.xml]: I know but the problem is that the project was with tabs, white spaces (2 spaces and 3 spaces) so IntelliJ has reordered in some way.

File: pom.xml (hunk 1 above)
Comment (jtnord) [inline pom.xml]: You did introduce this change - thats why making unrelated changes is a bad idea!

see a [whitespace ignoring diff](https://github.com/jenkinsci/jacoco-plugin/pull/63/files?w=1)

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 3 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: BCT? I am not sure that this is a breakwards incompatible change. AbstractBuild extends from Run so any AbstractBuild is also a Run.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 7]
```diff
@@ -34,348 +44,445 @@
...
-			        
-			        for (String dir : dirs) {
+
+    protected static FilePath[] resolveDirPaths(FilePath workspace, TaskListener listener, final String input) {
+        //final PrintStream logger = listener.getLogger();
+        FilePath[] directoryPaths = null;
+        try {
+            directoryPaths = workspace.act(new FilePath.FileCallable<FilePath[]>()
+            {
+                public void checkRoles(RoleChecker checker) throws SecurityException {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: this is the wrong thing to do.  use the correct sub type - see the javadoc for [FileCallable](http://javadoc.jenkins-ci.org/hudson/FilePath.FileCallable.html)

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 3 above)
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: The problem is that existing (already compiled) code may not be able to deal with the type change. I'm not sure in the case of a field though @jtnord , do you know for sure?

BCT is the [bytecode compatibility transformer](https://github.com/jenkinsci/bytecode-compatibility-transformer).

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 8]
```diff
@@ -59,15 +57,14 @@
 	/**
 	 * 
 	 * @param owner
-	 * @param rule
 	 * @param ratios
 	 *            The available coverage ratios in the report. Null is treated
 	 *            the same as an empty map.
 	 * @param thresholds
 	 */
-	public JacocoBuildAction(AbstractBuild<?,?> owner, Rule rule,
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Incompatible change.

File: src/main/java/hudson/plugins/jacoco/JacocoReportDir.java [hunk 9]
```diff
@@ -16,8 +16,8 @@
 public class JacocoReportDir {
     private final File root;
 
-    public JacocoReportDir(AbstractBuild<?,?> build) {
-        root = new File(build.getRootDir(), "jacoco");
+    public JacocoReportDir(File rootDir) {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoReportDir.java]: backwards incompatable change.

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java [hunk 10]
```diff
@@ -203,7 +204,7 @@ public Coverage getLineCoverage() {
 	/**
 	 * Gets the build object that owns the whole coverage report tree.
 	 */
-	public abstract AbstractBuild<?,?> getBuild();
+	public abstract Run<?,?> getBuild();
```
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: backwards incompatable change.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 7 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Ok, this was in original code, so I suppose I need to change it.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 11]
```diff
@@ -34,348 +44,445 @@
...
+    private String maximumInstructionCoverage;
+    private String maximumBranchCoverage;
+    private String maximumComplexityCoverage;
+    private String maximumLineCoverage;
+    private String maximumMethodCoverage;
+    private String maximumClassCoverage;
+    private boolean changeBuildStatus;
+
+    @DataBoundConstructor
+    public JacocoPublisher() {
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: The new constructor could just be empty since you have all the new `@DataboundSetters`. This would put an end to the sequence of `@Deprecated` constructors.

**Comment by jtnord** [comment]
so at least one :bug: on the `RoleSensitive` implementation on  `FileCallable`.

The backwards incompatable changes likely warrant another :bug: unless the maintainer is fine with this - and in this case you should go further with the changes and remove the public field (seeing as it could break pluginsanyway you may as well break them in a good way)

The whole formatting changes made this really hard to review as well - which is almost another bug in itself.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 12]
```diff
@@ -34,348 +44,445 @@
...
-	}
-	
-	protected static void saveCoverageReports(FilePath destFolder, FilePath sourceFolder) throws IOException, InterruptedException {
-		destFolder.mkdirs();
-		
-		sourceFolder.copyRecursiveTo(destFolder);
-	}
-	
-	protected static String resolveFilePaths(AbstractBuild<?, ?> build, BuildListener listener, String input) {
+    protected String resolveFilePaths(Run<?, ?> build, TaskListener listener, String input) {
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Incompatible change.

File: pom.xml (hunk 2 above)
Comment (jtnord) [inline pom.xml]: in which case a cleanup should have been ideally done before the changes.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 13]
```diff
@@ -34,348 +44,445 @@
...
-        try {
-        	healthReports = new JacocoHealthReportThresholds(Integer.parseInt(minimumClassCoverage), Integer.parseInt(maximumClassCoverage), Integer.parseInt(minimumMethodCoverage), Integer.parseInt(maximumMethodCoverage), Integer.parseInt(minimumLineCoverage), Integer.parseInt(maximumLineCoverage)
-        			,Integer.parseInt(minimumBranchCoverage), Integer.parseInt(maximumBranchCoverage), Integer.parseInt(minimumInstructionCoverage), Integer.parseInt(maximumInstructionCoverage), Integer.parseInt(minimumComplexityCoverage), Integer.parseInt(maximumComplexityCoverage));
-        } catch(NumberFormatException nfe) {
-        	healthReports = new JacocoHealthReportThresholds(0,0,0,0,0,0,0,0,0,0,0,0);
-        }		
-        
-        if ((execPattern==null) || (classPattern==null) || (sourcePattern==null)) {
-            if(build.getResult().isWorseThan(Result.UNSTABLE)) {
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Why are you keeping this one around?

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 3 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: It is, and it's not only the field but the `getOwner` method.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 14]
```diff
@@ -385,24 +492,35 @@ public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListene
...
+        try {
+            return healthReports = new JacocoHealthReportThresholds(Integer.parseInt(minimumClassCoverage), Integer.parseInt(maximumClassCoverage), Integer.parseInt(minimumMethodCoverage), Integer.parseInt(maximumMethodCoverage), Integer.parseInt(minimumLineCoverage), Integer.parseInt(maximumLineCoverage)
+                    , Integer.parseInt(minimumBranchCoverage), Integer.parseInt(maximumBranchCoverage), Integer.parseInt(minimumInstructionCoverage), Integer.parseInt(maximumInstructionCoverage), Integer.parseInt(minimumComplexityCoverage), Integer.parseInt(maximumComplexityCoverage));
+        } catch (NumberFormatException nfe) {
+            return healthReports = new JacocoHealthReportThresholds(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+        }
+    }
+
+
+    public static Result checkResult(JacocoBuildAction action) {
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Incompatible change.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 3 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: but a `Run` is not always an `AbstractBuild` so anyone expecting this to be an `AbstractBuild` may now blow up.
This is regardless of any binding that may or not have broken (I think the variable will actually be ok - but the type changing is not).  Why this needs to be `public` baffles me (but that is a different issue)

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java (hunk 10 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: But again with Workflow we need this, so you are suggesting https://github.com/jenkinsci/bytecode-compatibility-transformer#adapting-type

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 15]
```diff
@@ -201,12 +197,12 @@ public Object getTarget() {
 	}
 
 	@Override
-	public AbstractBuild<?,?> getBuild() {
+	public Run<?,?> getBuild() {
```
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Backward incompatibility here. You should use `@WithBridgeMethods`, [this](https://github.com/jenkinsci/analysis-core-plugin/blob/9f8f492d2931981f91a06a8eb0123fd7d3cd1a58/src/main/java/hudson/plugins/analysis/views/ConsoleDetail.java#L114-L128) is an example.

**Comment by lordofthejars** [comment]
Well I disagree about format changing, since it is only pom.xml, which in reality has been untouched and at least it fixes that something happens in the past that PR were accepted with different formats, but the other I agree with you

**Comment by daniel-beck** [comment]
There is no public plugin depending on this one (except the terrible DotCi starter pack), so the backward compatibility issue may be irrelevant.

**Comment by daniel-beck** [comment]
> I disagree about format changing, since it is only pom.xml

It's not. It's everywhere and it makes reviews terribly cumbersome.

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java (hunk 10 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: I would run a mile from the BCT.  What I am suggesting is you reach out to the plugin maintainer and ask them about breaking backwards API compatibility (and then if they agree do it properly :-) )

**Comment by lordofthejars** [comment]
Ok let me fix some of the problems and let's discuss with the maintainer if we want to add BCT or we can leave as is changing the attribute to private or package.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 16]
```diff
@@ -34,348 +44,445 @@
...
+        this.maximumClassCoverage = "0";
+        this.changeBuildStatus = false;
+    }
+
+    /**
      * Loads the configuration set by user.
      */
-    @DataBoundConstructor
+    @Deprecated
     public JacocoPublisher(String execPattern, String classPattern, String sourcePattern, String inclusionPattern, String exclusionPattern, String maximumInstructionCoverage, String maximumBranchCoverage
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: The next several hundred lines of diff are nothing but formatting changes. Also, it looks like there's some reordering of methods going on as well.

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java (hunk 10 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: the other way to do it is to deprecate this and introduce a new method `getRun()`  let `getBuild()` attempt to cast the Run to a AbstractBuild and throw an exception if its not possible - which should be ok as the code that currently calls this would only call it for a non workflow job...

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 4 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: It seems it's not used outside this plugin. It's an acceptable change IMO. Adding the `Utils.isOverridden` dance makes the code dirty, I would try to avoid it when possible.

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java (hunk 10 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: @amuniz points out the withBridgeMethods - but if there are no external users and the maintainer is ok with breaking the API - I would just break the API and reduce the scope wherever possible.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 15 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: However, it does not seem to be [used outside](https://github.com/search?utf8=%E2%9C%93&q=JacocoBuildAction+getOwner+user%3Ajenkinsci&type=Code&ref=searchresults), but who knows.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 17]
```diff
@@ -34,348 +44,445 @@
...
+    public void setInclusionPattern(String inclusionPattern) {
+        this.inclusionPattern = inclusionPattern;
+    }
 
+    @DataBoundSetter
+    public void setExclusionPattern(String exclusionPattern) {
+        this.exclusionPattern = exclusionPattern;
+    }
 
+    /** End Setters for Workflow **/
```
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Well, it's not for workflow. It's in general a good practice to use `@DataBoundSetter`s for fields that have a default value.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 12 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: It's protected. Acceptable IMO.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 18]
```diff
@@ -34,348 +44,445 @@
...
-        	healthReports = new JacocoHealthReportThresholds(0,0,0,0,0,0,0,0,0,0,0,0);
-        }		
-        
-        if ((execPattern==null) || (classPattern==null) || (sourcePattern==null)) {
-            if(build.getResult().isWorseThan(Result.UNSTABLE)) {
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+        return publishReports(build, build.getBuildVariables(), build.getRootDir(), build.getWorkspace(), launcher, listener);
+    }
+
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull Launcher launcher, @Nonnull TaskListener taskListener) throws InterruptedException, IOException {
```
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: `@Override`?

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 12 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: but the class is public and non-final.

File: src/main/java/hudson/plugins/jacoco/model/CoverageObject.java (hunk 10 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/model/CoverageObject.java]: :+1: to break API is possible (BTW it's useful to search in GitHub under user:jenkinsci before breaking it)

**Comment by amuniz** [comment]
Really difficult to read the diff. Sorry @recena I don't agree.

+1 to format the code, but in a separate PR please. I'm sure I missed some important changes because of formatting changes.

**Comment by amuniz** [comment]
@lordofthejars 

> but then Jenkins complains that this job has already created a directory and it is not executed

What's not executed? Did you try to use `@LocalData`?

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 5 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Plugin uses Java5, so I think that @Override cannot be used in interface. Also IntelliJ complains about it.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 6 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Plugin uses Java5, so I think that @Override cannot be used in interface. Also IntelliJ complains about it.

File: src/main/java/hudson/plugins/jacoco/JacocoReportDir.java (hunk 9 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoReportDir.java]: Same problem as usually Run is not an AbstractBuild

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 13 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: For compatibility with old method of JaCoCo

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 14 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Why it is only a helper method, I have substitute this call that was done in two places to a single one. It is an internal refactor.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 18 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: It is Java 5

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 5 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: the version of Jenkins (1.609.3) you are now targeting uses 1.6 - so you can fix that :-)

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 6 above)
Comment (jtnord) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: the version of Jenkins (1.609.3) you are now targeting uses 1.6 - so you can fix that :-)

**Comment by lordofthejars** [comment]
New Commit (number 2) fixes the `@Override` problem and updates Java version to 6. Also it creates a static nested class of `FileCallable`and adds an scope to a dependency.

**Comment by lordofthejars** [comment]
For me the change of AbstractBuild -> Run is ok and it should not be a problem at all. As all of you suggest I prefer to not start using BCT

File: pom.xml (hunk 1 above)
Comment (lordofthejars) [inline pom.xml]: It is ok with no scope since it is used in one class. In previous Jenkins dependency (1.424.6) this artifact was resolved with transitivity. Now that we have moved the minimum Jenkins version to 1.609, this dependency is not there anymore and must be added specifically.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 17 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: this is what I have done right?

Comment (amuniz) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Yes... I mean, you can remove this comment - it's not accurate and it could lead to confusion.

**Comment by jglick** [comment]
With the whitespace changes this is unreviewable. Let us know when you can prepare a minimal diff.

**Comment by lordofthejars** [comment]
Well the only thing that I can do is discard the change and rework it again.

**Comment by lordofthejars** [comment]
@jglick have you tried passing w=1 to the url? https://github.com/jenkinsci/jacoco-plugin/pull/63/files?w=1 to avoid white space diffs? Next time I will not do automatic format but at least if you can review it with this option I wouldn't have to redo the change.

**Comment by centic9** [comment]
But reviewing without whitespaces will still leave the PR unmergeable due to the huge number of unrelated changes! I have compared the changes locally and created PR #66 with the reduced changes to allow us to make progress on this one...  

I suggest you base further work on that branch.

This PR is therefore obsoleted now by #66 .

**Comment by lordofthejars** [comment]
@centic9 Thanks, I started the same last evening.
